### DISCUSSION TITLE
关于xmake使用libc_nano问题

### DISCUSSION THREAD

**Discussion by PeakRacing** [xmake-io/xmake#1930] [Q&A]
您好，我在使用xmake编译嵌入式平台，想使用gcc的libc_nano
![image](https://user-images.githubusercontent.com/24759100/137063267-4ccdbedb-015f-40fb-ac60-64365de5308b.png)
添加了--specs=nano.specs，但是编译固件大小还是大，查看map未使用libc_nano
![image](https://user-images.githubusercontent.com/24759100/137063821-ef0d6e81-014e-4ac9-a66b-aa5f55d80e8a.png)
使用makefile编译正常
请问是我哪里使用错误么？
xmake版本2.5.8
gcc使用gcc-arm-none-eabi-2021.07版本10.3

**Comment by waruqi** [comment]
自己 执行 xmake -v 看下实际编译命令，确认自己设置的 flags 有没有设置到 gcc 上，如果确实已经设置了，那说明 xmake 没问题。。如果还是没生效，你要去 gcc 那边反馈

**Comment by waruqi** [comment]
另外，确认下文件的相对路径问题

**Comment by PeakRacing** [comment]
> 另外，确认下文件的相对路径问题

你好之前查过xmake -v有标志
![image](https://user-images.githubusercontent.com/24759100/137065067-52f134c3-16b8-4716-9d98-68909836109a.png)
请问相对路径注意什么？
![image](https://user-images.githubusercontent.com/24759100/137065157-976434ef-d4e0-4d38-ba90-b886d33902e5.png)
现在是这样的，就是不使用lib_nano

**Comment by waruqi** [comment]
你传入 ldflags的文件路径是字符串 不会走 add_files 这种相对于当前脚本的路径转换，xmake 默认构建目录在根目录，自己控制下传入的路径对 gcc 来说是否能正常访问。。如果路径确认也有效，也有flags传入gcc 那就跟 xmake 一点关系都没有。。

其他的就是gcc自身的事情了，自己想办法解决。。

**Comment by waruqi** [comment]
还是不行么 可以切到 `$(projectdir)/xxx.ld` 走绝对路径

另外不要一长串全塞一个 add_ldflags 每个带空格的flags项，单独设置

add_ldflags("-Wl,aaa", "-Wl,bbb", "-Wl,ccc")

**Comment by PeakRacing** [comment]
> 还是不行么 可以切到 `$(projectdir)/xxx.ld` 走绝对路径
> 
> 另外不要一长串全塞一个 add_ldflags 每个带空格的flags项，单独设置
> 
> add_ldflags("-Wl,aaa", "-Wl,bbb", "-Wl,ccc")

您好
![image](https://user-images.githubusercontent.com/24759100/137506445-88bd36d6-0b62-473b-8db4-f04b2a45f633.png)
![image](https://user-images.githubusercontent.com/24759100/137506509-1c30c0f1-6954-4849-88c2-ada845f63420.png)
已经改成这样了，状况还是和之前一样
![image](https://user-images.githubusercontent.com/24759100/137505838-2787a54a-a6a4-448f-8c8d-9a8dde1af37d.png)
map文件看还是libc而不是libc_nano
![image](https://user-images.githubusercontent.com/24759100/137505998-8231346e-c0ed-4dc9-a695-f8119900ac00.png)
![image](https://user-images.githubusercontent.com/24759100/137506080-ae955035-09ad-4b61-bf50-645d13d32c28.png)
Makefile是使用的libc_nano

**Comment by waruqi** [comment]
xmake -v 看下实际的命令参数  跟 makefile 的自己对比下差异

**Comment by github0null** [comment]
切换libc是在链接时进行的，因此 `--specs=nosys.specs --specs=nano.specs` 要加在 **LDFLAGS** 参数里，加到 **CFLAGS** 肯定没用的

**Comment by PeakRacing** [comment]
> 切换libc是在链接时进行的，因此 `--specs=nosys.specs --specs=nano.specs` 要加在 **LDFLAGS** 参数里，加到 **CFLAGS** 肯定没用的

 LDFLAGS里加了--specs=nano.specs，情况一样，最后在 LDFLAGS里直接加-lc_nano可以了

**Comment by github0null** [comment]
嗯嗯

一般这个 `--specs=nosys.specs --specs=nano.specs` 这两个是一对，一起指定，就不用再加 `-lc_nano` 了

**Comment by PeakRacing** [comment]
> 嗯嗯
> 
> 一般这个 `--specs=nosys.specs --specs=nano.specs` 这两个是一对，一起指定，就不用再加 `-lc_nano` 了

不知道为啥加--specs=nosys.specs --specs=nano.specs不好使，不过-lc_nano好使就这么用了，哈哈

**Comment by waruqi** [comment]
既然可以了，就先这样吧。
